From 79bf03101679a970c84895e7a9625b8a205c4f67 Mon Sep 17 00:00:00 2001
From: Alexander Winkowski <dereference23@outlook.com>
Date: Sat, 4 Feb 2023 14:50:57 +0000
Subject: [PATCH] net: Don't reduce TTL/HL

Some mobile carriers compare TTL/HL values to restrict
tethering. Just skip the decrement to make it harder for them.

Signed-off-by: Alexander Winkowski <dereference23@outlook.com>
---
 net/ipv4/ip_forward.c | 11 ++++++-----
 net/ipv6/ip6_output.c |  8 ++++----
 2 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/net/ipv4/ip_forward.c b/net/ipv4/ip_forward.c
index 66fac1216d467..5ce45605fa744 100644
--- a/net/ipv4/ip_forward.c
+++ b/net/ipv4/ip_forward.c
@@ -111,7 +111,7 @@ int ip_forward(struct sk_buff *skb)
 
 	skb_forward_csum(skb);
 	net = dev_net(skb->dev);
-
+#if 0
 	/*
 	 *	According to the RFC, we must first decrease the TTL field. If
 	 *	that reaches zero, we must reply an ICMP control message telling
@@ -119,7 +119,7 @@ int ip_forward(struct sk_buff *skb)
 	 */
 	if (ip_hdr(skb)->ttl <= 1)
 		goto too_many_hops;
-
+#endif
 	if (!xfrm4_route_forward(skb)) {
 		SKB_DR_SET(reason, XFRM_POLICY);
 		goto drop;
@@ -144,10 +144,10 @@ int ip_forward(struct sk_buff *skb)
 	if (skb_cow(skb, LL_RESERVED_SPACE(rt->dst.dev)+rt->dst.header_len))
 		goto drop;
 	iph = ip_hdr(skb);
-
+#if 0
 	/* Decrease ttl after skb cow done */
 	ip_decrease_ttl(iph);
-
+#endif
 	/*
 	 *	We now generate an ICMP HOST REDIRECT giving the route
 	 *	we calculated.
@@ -169,12 +169,13 @@ int ip_forward(struct sk_buff *skb)
 	 */
 	 icmp_send(skb, ICMP_DEST_UNREACH, ICMP_SR_FAILED, 0);
 	 goto drop;
-
+#if 0
 too_many_hops:
 	/* Tell the sender its packet died... */
 	__IP_INC_STATS(net, IPSTATS_MIB_INHDRERRORS);
 	icmp_send(skb, ICMP_TIME_EXCEEDED, ICMP_EXC_TTL, 0);
 	SKB_DR_SET(reason, IP_INHDR);
+#endif
 drop:
 	kfree_skb_reason(skb, reason);
 	return NET_RX_DROP;
diff --git a/net/ipv6/ip6_output.c b/net/ipv6/ip6_output.c
index cd89a2b35dfb5..760e8e6a15703 100644
--- a/net/ipv6/ip6_output.c
+++ b/net/ipv6/ip6_output.c
@@ -546,7 +546,7 @@ int ip6_forward(struct sk_buff *skb)
 		if (ip6_call_ra_chain(skb, ntohs(opt->ra)))
 			return 0;
 	}
-
+#if 0
 	/*
 	 *	check and decrement ttl
 	 */
@@ -557,7 +557,7 @@ int ip6_forward(struct sk_buff *skb)
 		kfree_skb_reason(skb, SKB_DROP_REASON_IP_INHDR);
 		return -ETIMEDOUT;
 	}
-
+#endif
 	/* XXX: idev->cnf.proxy_ndp? */
 	if (net->ipv6.devconf_all->proxy_ndp &&
 	    pneigh_lookup(&nd_tbl, net, &hdr->daddr, skb->dev, 0)) {
@@ -659,9 +659,9 @@ int ip6_forward(struct sk_buff *skb)
 	hdr = ipv6_hdr(skb);
 
 	/* Mangling hops number delayed to point after skb COW */
-
+#if 0
 	hdr->hop_limit--;
-
+#endif
 	return NF_HOOK(NFPROTO_IPV6, NF_INET_FORWARD,
 		       net, NULL, skb, skb->dev, dst->dev,
 		       ip6_forward_finish);
